pulsar-broker/src/main/java/org/apache/pulsar/compaction/StrategicTwoPhaseCompactor.java

```diff
@@ -443,32 +421,48 @@ private <T> void phaseTwoLoop(String topic, Iterator<Message<T>> reader,
 
     <T> CompletableFuture<Boolean> addToCompactedLedger(
             LedgerHandle lh, Message<T> m, String topic, Semaphore outstanding) {
+        if (m == null) {
+            return flushBatchMessage(lh, topic, outstanding);
+        }
+        if (batchMessageContainer.haveEnoughSpace((MessageImpl<?>) m)) {
+            if (batchMessageContainer.add((MessageImpl<?>) m, null)) {
+                return flushBatchMessage(lh, topic, outstanding);
+            }
+            return CompletableFuture.completedFuture(false);
+        }
+        CompletableFuture<Boolean> f = flushBatchMessage(lh, topic, outstanding);
+        if (batchMessageContainer.add((MessageImpl<?>) m, null)) {
+            return flushBatchMessage(lh, topic, outstanding).thenCombine(f, (a, b) -> a && b);
+        } else {
+            return f;
+        }
+    }
+
+    private CompletableFuture<Boolean> flushBatchMessage(LedgerHandle lh, String topic,
+                                                         Semaphore outstanding) {
         CompletableFuture<Boolean> bkf = new CompletableFuture<>();
-        if (m == null || batchMessageContainer.add((MessageImpl<?>) m, null)) {
-            if (batchMessageContainer.getNumMessagesInBatch() > 0) {
-                try {
-                    ByteBuf serialized = batchMessageContainer.toByteBuf();
-                    outstanding.acquire();
-                    mxBean.addCompactionWriteOp(topic, serialized.readableBytes());
-                    long start = System.nanoTime();
-                    lh.asyncAddEntry(serialized,
-                            (rc, ledger, eid, ctx) -> {
-                                outstanding.release();
-                                mxBean.addCompactionLatencyOp(topic, System.nanoTime() - start, TimeUnit.NANOSECONDS);
-                                if (rc != BKException.Code.OK) {
-                                    bkf.completeExceptionally(BKException.create(rc));
-                                } else {
-                                    bkf.complete(true);
-                                }
-                            }, null);
+        if (batchMessageContainer.getNumMessagesInBatch() > 0) {
```

Comment:
I see you already made much changes to `addToCompactedLedger`, for the refactored code, you can just use early return.

The previous code:

```java
        if (m == null || batchMessageContainer.add((MessageImpl<?>) m, null)) {
            if (batchMessageContainer.getNumMessagesInBatch() > 0) {
                try {
                    /* ... */
                } catch (Throwable t) {
                    log.error("Failed to add entry", t);
                    batchMessageContainer.discard((Exception) t);
                    return FutureUtil.failedFuture(t);
                }
            } else {
                bkf.complete(false);
            }
        } else {
            bkf.complete(false);
        }
```

After applying early return idiom:

```java
        if (m != null && !batchMessageContainer.add((MessageImpl<?>) m, null)) {
            return CompletableFuture.completedFuture(false);
        }
        if (batchMessageContainer.getNumMessagesInBatch() <= 0) {
            return CompletableFuture.completedFuture(false);
        }
        // try-catch...
```